### ISSUE TITLE
Terraform crash (mostly AWS modules) during plan/apply/destroy

### ISSUE THREAD

**Issue by nbertram** [hashicorp/terraform#19357]
---
name: Bug report
about: Let us know about an unexpected error, a crash, or an incorrect behavior.

---

### Terraform Version

```
Terraform v0.11.10
+ provider.aws v1.43.2
+ provider.template v1.0.0

```

### Terraform Configuration Files

Not sure exactly which part has triggered this, though suspect something around the subnet/route associations in subnets.tf since that was the most recent set of changes...

Nothing very sensitive in the config:

https://seafile.catalyst.net.nz/f/dc068f654eb849469254/?dl=1

### Debug Output

Trace (encrypted, gzipped):
https://seafile.catalyst.net.nz/f/c70022c9f65f49359e46/?dl=1

### Crash Output

https://seafile.catalyst.net.nz/f/63d1b22cbf7446ef8ba9/?dl=1

### Expected Behavior

Terraform plan should have presented a change to aws_route53_record.proxy I think.

### Actual Behavior

Crashed as above. Not easy to ascertain the cause.

### Steps to Reproduce

terraform init
terraform plan

It doesn't crash unless the tfstate is present (plan to create fresh doesn't seem to crash), so it's likely the state is important.

Encrypted tfstate is at https://seafile.catalyst.net.nz/f/162df7cb37ba47e88f79/?dl=1

### Additional Context

Terraform is wrapped in Terragrunt v0.16.10, though not currently using its functionality. It crashes with or without Terragrunt.

### References

Nothing I'm aware of.

**Comment by bflad** [comment]
Reopening here as this looks like a panic in the provider SDK:

```
2018-11-13T16:45:21.525+1300 [DEBUG] plugin.terraform-provider-aws_v1.43.2_x4: panic: Error reading level config: '' expected type 'string', got unconvertible type '[]interface {}'
2018-11-13T16:45:21.525+1300 [DEBUG] plugin.terraform-provider-aws_v1.43.2_x4: 
2018-11-13T16:45:21.525+1300 [DEBUG] plugin.terraform-provider-aws_v1.43.2_x4: goroutine 1444 [running]:
2018-11-13T16:45:21.525+1300 [DEBUG] plugin.terraform-provider-aws_v1.43.2_x4: github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/schema.(*ResourceData).get(0xc000379180, 0xc00015ef60, 0x1, 0x1, 0xc00015ef12, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
2018-11-13T16:45:21.525+1300 [DEBUG] plugin.terraform-provider-aws_v1.43.2_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/schema/resource_data.go:509 +0x449
2018-11-13T16:45:21.525+1300 [DEBUG] plugin.terraform-provider-aws_v1.43.2_x4: github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/schema.(*ResourceData).getChange(0xc000379180, 0x31ec824, 0x2, 0xc000391201, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
2018-11-13T16:45:21.525+1300 [DEBUG] plugin.terraform-provider-aws_v1.43.2_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/schema/resource_data.go:480 +0x122
2018-11-13T16:45:21.525+1300 [DEBUG] plugin.terraform-provider-aws_v1.43.2_x4: github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/schema.(*ResourceData).diffChange(0xc000379180, 0x31ec824, 0x2, 0xe, 0xc00039f5a8, 0x0, 0x0, 0x0)
2018-11-13T16:45:21.525+1300 [DEBUG] plugin.terraform-provider-aws_v1.43.2_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/schema/resource_data.go:457 +0x97
2018-11-13T16:45:21.525+1300 [DEBUG] plugin.terraform-provider-aws_v1.43.2_x4: github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/schema.schemaMap.diffString(0xc0004d9ec0, 0x31ec824, 0x2, 0xc0004c83c0, 0xc000af5640, 0x3513840, 0xc000379180, 0xc000569600, 0x0, 0x0)
2018-11-13T16:45:21.525+1300 [DEBUG] plugin.terraform-provider-aws_v1.43.2_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/schema/schema.go:1158 +0xa0
2018-11-13T16:45:21.525+1300 [DEBUG] plugin.terraform-provider-aws_v1.43.2_x4: github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/schema.schemaMap.diff(0xc0004d9ec0, 0x31ec824, 0x2, 0xc0004c83c0, 0xc0002f0380, 0x3513840, 0xc000379180, 0xa528cb028b69e800, 0x0, 0x0)
2018-11-13T16:45:21.525+1300 [DEBUG] plugin.terraform-provider-aws_v1.43.2_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/schema/schema.go:774 +0x436
2018-11-13T16:45:21.525+1300 [DEBUG] plugin.terraform-provider-aws_v1.43.2_x4: github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/schema.schemaMap.Diff(0xc0004d9ec0, 0x0, 0xc000325e00, 0x0, 0x2c741e0, 0xc000230300, 0x71374491428a2f98, 0xe9b5dba5b5c0fbcf, 0x411d51)
2018-11-13T16:45:21.525+1300 [DEBUG] plugin.terraform-provider-aws_v1.43.2_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/schema/schema.go:424 +0x206
2018-11-13T16:45:21.525+1300 [DEBUG] plugin.terraform-provider-aws_v1.43.2_x4: github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/schema.(*Resource).Diff(0xc0001bf6c0, 0x0, 0xc000325e00, 0x2c741e0, 0xc000230300, 0xc000569b01, 0x4bfe7c, 0x27519c0)
2018-11-13T16:45:21.525+1300 [DEBUG] plugin.terraform-provider-aws_v1.43.2_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/schema/resource.go:250 +0x17c
2018-11-13T16:45:21.525+1300 [DEBUG] plugin.terraform-provider-aws_v1.43.2_x4: github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/schema.(*Provider).ReadDataDiff(0xc0008159d0, 0xc00058a000, 0xc000325e00, 0xc00036a380, 0x53380d134d2c6dfc, 0x7f9f2760e000)
2018-11-13T16:45:21.525+1300 [DEBUG] plugin.terraform-provider-aws_v1.43.2_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/schema/provider.go:413 +0x9b
2018-11-13T16:45:21.525+1300 [DEBUG] plugin.terraform-provider-aws_v1.43.2_x4: github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/plugin.(*ResourceProviderServer).ReadDataDiff(0xc000849360, 0xc00015e4e0, 0xc00015e650, 0x0, 0x0)
2018-11-13T16:45:21.525+1300 [DEBUG] plugin.terraform-provider-aws_v1.43.2_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/plugin/resource_provider.go:593 +0x4e
2018-11-13T16:45:21.525+1300 [DEBUG] plugin.terraform-provider-aws_v1.43.2_x4: reflect.Value.call(0xc00038d020, 0xc000172060, 0x13, 0x31ed258, 0x4, 0xc000569f18, 0x3, 0x3, 0xc000a1d9c0, 0xc000142300, ...)
2018-11-13T16:45:21.525+1300 [DEBUG] plugin.terraform-provider-aws_v1.43.2_x4: 	/opt/goenv/versions/1.11.1/src/reflect/value.go:447 +0x449
2018-11-13T16:45:21.525+1300 [DEBUG] plugin.terraform-provider-aws_v1.43.2_x4: reflect.Value.Call(0xc00038d020, 0xc000172060, 0x13, 0xc000b74f18, 0x3, 0x3, 0x5d464b, 0xc000b74f28, 0x429732)
2018-11-13T16:45:21.525+1300 [DEBUG] plugin.terraform-provider-aws_v1.43.2_x4: 	/opt/goenv/versions/1.11.1/src/reflect/value.go:308 +0xa4
2018-11-13T16:45:21.525+1300 [DEBUG] plugin.terraform-provider-aws_v1.43.2_x4: net/rpc.(*service).call(0xc0008aa040, 0xc0004e60f0, 0xc00050e050, 0xc00050e060, 0xc0001b6500, 0xc000a2c900, 0x2751980, 0xc00015e4e0, 0x16, 0x27519c0, ...)
2018-11-13T16:45:21.525+1300 [DEBUG] plugin.terraform-provider-aws_v1.43.2_x4: 	/opt/goenv/versions/1.11.1/src/net/rpc/server.go:384 +0x14e
2018-11-13T16:45:21.525+1300 [DEBUG] plugin.terraform-provider-aws_v1.43.2_x4: created by net/rpc.(*Server).ServeCodec
2018-11-13T16:45:21.525+1300 [DEBUG] plugin.terraform-provider-aws_v1.43.2_x4: 	/opt/goenv/versions/1.11.1/src/net/rpc/server.go:481 +0x47e
```

**Comment by bflad** [comment]
Hello again, @nbertram 👋 I'll let the maintainers more familiar with this certain code discuss any potential fixes or enhancements there to prevent this situation (I believe this might actually be resolved or at least return a better error message in the Terraform 0.12 alphas), but I'm wondering if we can't narrow down to where in your configuration this is occurring.

The message `panic: Error reading level config: '' expected type 'string', got unconvertible type '[]interface {}'` seems to be indicating that your configuration might have a list of items in a place where only a single string is expected. Very briefly looking through your code I did find this place where that might be an issue:

```hcl
# ecs_cluster/security_groups.tf
resource "aws_security_group" "proxy" {
  # ... other configuration ...

  ingress {
    # NLBs have weird security group settings that mean that although the traffic
    # is travelling through an NLB, the source address of the traffic is preserved.
    # We therefore allow port 3128 from places inside this cluster and additionally the
    # support subnet in case it wasnts to use this proxy.
    from_port = 3128
    to_port   = 3128
    protocol  = "tcp"
    cidr_blocks = [
      "${data.aws_subnet.ecs-subnet.*.cidr_block}",
      "${data.aws_subnet.support.cidr_block}"
    ]
    ipv6_cidr_blocks = [
      "${data.aws_subnet.ecs-subnet.*.ipv6_cidr_block}",
      "${data.aws_subnet.support.ipv6_cidr_block}"
    ]
  }
}
```

I'm wondering if the merging of the list of `data.aws_subnet.ecs-subnet.*.cidr_block` with the string of `data.aws_subnet.support.cidr_block` might be a culprit here (same for the `ipv6_cidr_block`). Could you try hardcoding the individual data source references:

```hcl
    cidr_blocks = [
      "${data.aws_subnet.ecs-subnet.0.cidr_block}",
      "${data.aws_subnet.ecs-subnet.1.cidr_block}",
      "${data.aws_subnet.ecs-subnet.2.cidr_block}",
      "${data.aws_subnet.support.cidr_block}"
    ]
    ipv6_cidr_blocks = [
      "${data.aws_subnet.ecs-subnet.0.ipv6_cidr_block}",
      "${data.aws_subnet.ecs-subnet.1.ipv6_cidr_block}",
      "${data.aws_subnet.ecs-subnet.2.ipv6_cidr_block}",
      "${data.aws_subnet.support.ipv6_cidr_block}"
    ]
```

Or wrapping the entire list with `list()` (apologies the syntax might be slightly off here):

```hcl
    cidr_blocks = ["${list(
      data.aws_subnet.ecs-subnet.*.cidr_block,
      data.aws_subnet.support.cidr_block,
    )}"]
    ipv6_cidr_blocks = ["${list(
      data.aws_subnet.ecs-subnet.*.ipv6_cidr_block,
      data.aws_subnet.support.ipv6_cidr_block,
    )}"]
```

To see if they help?

**Comment by apparentlymart** [comment]
The root cause here seems to be the same as #16902, which is already fixed in master and ready to be included in the forthcoming v0.12.0 release.

The expressions that @bflad indicated to indeed seem like the trigger for the problem, since they are producing a list containing another list rather than a flat list of strings. I think @bflad's second suggestion has the same problem -- it passes a list and then a string to `list`, which is not allowed -- but here's a different expression which I think will achieve the desired effect of a single flat list of strings:

```hcl
  cidr_blocks = "${concat(data.aws_subnet.ecs-subnet.*.cidr_block, list(data.aws_subnet.support.cidr_block))}"
```

The `concat` function takes two lists and concatenates them together to produce a single list. The result of the splat expression is already a list, but the single CIDR block string from the `support` subnet needs to be wrapped in a list itself as shown above.

In the forthcoming v0.12.0 release a slightly more readable form of this will become possible:

```hcl
  # Not supported in v0.11; coming in v0.12
  cidr_blocks = concat(
    data.aws_subnet.ecs-subnet.*.cidr_block,
    [data.aws_subnet.support.cidr_block],
  )
```

**Comment by nbertram** [comment]
@apparentlymart yes I'll admit that syntax did seem problematic, though switching it to either the concat() or a fixed list (or removing the entire resource) hasn't stopped the crash. I wonder if I've done the same thing somewhere else, though it's not obvious if I have.

It might just be easier to try a 0.12 build and see if it can point out the issues for me.

**Comment by nbertram** [comment]
OK found it. It was a typo, though a dangerous one.

```diff
 data "aws_subnet" "elb-subnet" {
   count = "3"
-  id = "${var.elb_subnet_ids}"
+  id = "${var.elb_subnet_ids[count.index]}"
 }
```

Trying to pass a list of subnet IDs to a string attribute. I imagine this makes it an issue with the AWS provider again?

**Comment by apparentlymart** [comment]
Hi @nbertram! Thanks for following up.

We've seen that crash with assigning a list to the subnet id before in #16902, and I closed it after verifying that it was fixed in on of the v0.12.0 alpha builds.

The fix that we've put in place is to actually validate that in the Terraform Core layer. The fact that the missing validation caused a crash in the provider SDK is also concerning since it points to a lack of robustness in the SDK, but addressing that will probably wait until a later planned effort to revamp the provider SDK to support some of the new protocol capabilities that 0.12 has introduced, which we're postponing for right now because we need to have a period where new provider releases support both v0.11 and v0.12 to allow for a gradual migration path.

Since the specific situation that was causing the crash was already verified fixed elsewhere, I'm going to close this out now. Thanks for the investigation work to track that down!
